# Cas d'usage n°1

Après récupération d'un jeu de données quelconque (ici les industries classées pour la protection de l'environnement = ICPE de la région Hauts-de-France), l'objectif est de retrouver les codes communes manquants.

## Import des modules

In [1]:
import os
import pandas as pd
from requests_cache import CachedSession
from tqdm import tqdm

from french_cities import find_city

## Constitution du jeu de données : récupération des ICPE de la région Hauts-de-France

In [2]:
code_region = "32"
page_size = 1000

s = CachedSession()
r = s.get(
    "https://georisques.gouv.fr/api/v1/installations_classees",
    params={"page": "1", "page_size": page_size, "region": code_region},
)
number_pages = r.json()["total_pages"]
for x in tqdm(range(number_pages), desc="querying georisques", leave=False):
    try:
        data
    except NameError:
        data = []
    else:
        r = s.get(
            "https://georisques.gouv.fr/api/v1/installations_classees",
            params={
                "page": x + 1,
                "page_size": page_size,
                "region": code_region,
            },
        )
    finally:
        data += r.json()["data"]
    if not r.json()["next"]:
        break
data = pd.DataFrame(data)

## Analyse du jeu de données obtenu

In [3]:
data

raisonSociale                       adresse1 codePostal  \
0              SCEA FERME LABALETTE             Lieu-dit Le Dièvre      62860   
1      ISDI de Saint-Laurent-Blangy       Zone des Trois Fontaines      62223   
2            RECYCL'ELECTRONIC SARL   2 Bis Rue <charles Duquesnoy      62270   
3           ISDI -TCPA (PATINIER A)                 rue de la Gare      62470   
4                          LAV'ALIM    Zone Industrielle Le Royeux      02430   
...                             ...                            ...        ...   
12360                          LIDL              1 B RUE DE CALAIS      59820   
12361                       NORAUTO  ZAL n°1 - Rue Bernard Chochoy      62800   
12362                 CAPART Julien        Parcelle cadastrée ZN 4      80340   
12363            GCA SUPPLY PACKING          78 RUE HENRI BARBUSSE      59224   
12364         Mairie de Toutencourt                   3 Rue Longue      80560   

      codeInsee               commune  longitude   latitude  bovins  porcs  \
0         62739    Sains-lès-Marquion   3.083774  50.176359   False  False   
1         62753  Saint-Laurent-Blangy   2.809392  50.303387   False  False   
2         62361               Frévent   2.293696  50.267633   False  False   
3         62197    Camblain-Châtelain   2.429339  50.478130   False  False   
4         02340                Gauchy   3.293073  49.822421   False  False   
...         ...                   ...        ...        ...     ...    ...   
12360     59273            GRAVELINES   2.118673  50.986002   False  False   
12361     62510                Liévin   2.773560  50.422542   False  False   
12362     80432            Herleville   2.739324  49.874223   False  False   
12363     59589                THIANT   3.442195  50.307796   False  False   
12364     80766           Toutencourt   2.458095  50.035682   False  False   

       volailles  ...  serviceAIOT          regime  \
0          False  ...    DREAL HdF  Enregistrement   
1          False  ...    DREAL HdF        Non ICPE   
2          False  ...    DREAL HdF  Autres régimes   
3          False  ...    DREAL HdF        Non ICPE   
4          False  ...    DREAL HdF  Autres régimes   
...          ...  ...          ...             ...   
12360      False  ...    DREAL HdF        Non ICPE   
12361      False  ...    DREAL HdF        Non ICPE   
12362      False  ...    DREAL HdF        Non ICPE   
12363      False  ...    DREAL HdF        Non ICPE   
12364      False  ...    DREAL HdF        Non ICPE   

                                               rubriques  \
0      [{'numeroRubrique': '2760', 'nature': 'Install...   
1                                                     []   
2                                                     []   
3                                                     []   
4                                                     []   
...                                                  ...   
12360                                                 []   
12361                                                 []   
12362                                                 []   
12363                                                 []   
12364                                                 []   

                                             inspections  \
0      [{'dateInspection': '2024-06-28', 'fichierInsp...   
1                     [{'dateInspection': '2019-02-04'}]   
2                     [{'dateInspection': '2015-09-29'}]   
3                     [{'dateInspection': '2016-12-06'}]   
4                     [{'dateInspection': '2020-02-24'}]   
...                                                  ...   
12360                 [{'dateInspection': '2024-12-02'}]   
12361                 [{'dateInspection': '2024-12-16'}]   
12362                 [{'dateInspection': '2024-12-09'}]   
12363                 [{'dateInspection': '2024-12-10'}]   
12364                 [{'dateInspection': '2024-12-09'}]   

                     

## ICPE dépourvues de codes commune INSEE

In [4]:
data.codeInsee.isnull().value_counts().to_frame("Nombre de codes manquants :")

Nombre de codes manquants :
codeInsee                             
False                            12112
True                               253

##### A date du 14/02/2025, le jeu de données est constitué de 12365 lignes et 31 colonnes, dont 253 codes communes INSEE manquants. Pourtant, d'autres champs
sont disponibles et exploitables pour retrouver les communes manquantes :
* des champs adresse ;
* un libellé de commune ;
* des coordonnées géographiques et référentiel de projection ;
* un code postal.

## Complétion des données avec `french-cities`

### Extraction des données manquantes

In [5]:
missing = data[data.codeInsee.isnull()].copy()

Au besoin, vérifier que le système de projection des coordonnées est en EPSG 2154 (lorsqu'il est connu) :

In [6]:
missing[["systemeCoordonneesAIOT"]].fillna("est manquant").value_counts().to_frame()

count
systemeCoordonneesAIOT       
est manquant              236
2154                       17

### Concaténation des adresses :

In [7]:
cols = [f"adresse{x}" for x in range(1, 4)]
addresses = (
    missing[cols[0]]
    .str.cat(missing[cols[1:]], sep=" ", na_rep="")
    .str.replace(" +", " ", regex=True)
    .str.strip(" ")
)
missing["adresse"] = addresses

### Recherche des communes manquantes à l'aide de `french-cities` :

In [8]:
filled = find_city(
    missing,
    year="last",
    x="coordonneeXAIOT",
    y="coordonneeYAIOT",
    epsg=2154,
    city="commune",
    dep=False,
    address="adresse",
    postcode="codePostal",
    use_nominatim_backend=False,
    field_output="newCodeInsee",
)
filled

No data found !area for ultra-marine territories:   0%|                                                                              | 0/8 [00:00<?, ?it/s]
No data found !
No data found !
No data found !
No data found !area for ultra-marine territories:  50%|███████████████████████████████████                                   | 4/8 [00:02<00:02,  1.39it/s]
No data found !
                                                                                                                                                           

raisonSociale                    adresse1  \
16                 SARL FINANCIERE VARET  Lieu-dit "les Huit Cannes"   
502                          GSMC Market     ZAC DU PONT SANS PAREIL   
2782                      VIOLET Bernard              La Grande Cote   
3675                           ALTEA MSO              ZAC du Plateau   
3751          LES FRUITS ROUGES AISNE SA            1 rue Jean Bodin   
...                                  ...                         ...   
12148                  CRAPET Jean-Louis          le fond du Marquoi   
12166  PRESSING DES PORTES DE L'AVESNOIS         RTE DE VALENCIENNES   
12167                          VAL PRESS             72 AV DE DENAIN   
12212                            NORBLOC         50 route de Tournai   
12217                        AUTO SELECT       394 B RTE DE FEIGNIES   

      codePostal codeInsee              commune   longitude   latitude  \
16         62118       NaN              Fampoux   -2.279526  45.502618   
502        62610       NaN       BOIS EN ARDRES         NaN        NaN   
2782       60180       NaN      Nogent-sur-Oise  114.936723 -89.574265   
3675       02200       NaN               Ploisy         NaN        NaN   
3751       02000       NaN                 Laon   -1.363081  -5.983856   
...          ...       ...                  ...         ...        ...   
12148      62550       NaN  Bailleul-lès-Pernes   -5.525510  42.559454   
12166      59530       NaN           LE QUESNOY   -0.001143   0.001046   
12167      59300       NaN         VALENCIENNES   -0.001105   0.001474   
12212      59226       NaN             Lecelles   -0.001220  -0.001001   
12217      59600       NaN             MAUBEUGE   -0.002286   0.000888   

       bovins  porcs  volailles  ...  \
16      False  False      False  ...   
502     False  False      False  ...   
2782    False  False      False  ...   
3675    False  False      False  ...   
3751    False  False      False  ...   
...       ...    ...        ...  ...   
12148   False  False      False  ...   
12166   False  False      False  ...   
12167   False  False      False  ...   
12212   False  False      False  ...   
12217   False  False      False  ...   

                                               rubriques  \
16     [{'numeroRubrique': '2760', 'nature': 'Install...   
502                                                   []   
2782   [{'numeroRubrique': '2510', 'nature': 'Carrièr...   
3675                                                  []   
3751   [{'numeroRubrique': '1511', 'nature': 'Entrepô...   
...                                                  ...   
12148                                                 []   
12166                                                 []   
12167                                                 []   
12212                                                 []   
12217                                                 []   

                                             inspections  \
16     [{'dateInspection': '2024-06-25', 'fichierInsp...   
502                   [{'dateInspection': '2017-11-10'}]   
2782   [{'dateInspection': '2024-07-25', 'fichierInsp...   
3675                                                  []   
3751   [{'dateInspection': '2022-09-21', 'fichierInsp...   
...                                                  ...   
12148  [{'dateInspection': '2023-10-06', 'fichierInsp...   
12166  [{'dateInspection': '2023-10-11', 'fichierInsp...   
12167  [{'dateInspection': '2023-10-11', 'fichierInsp...   
12212  [{'dateInspection': '2024-02-16', 'fichierInsp...   
12217  [{'dateInspection': '2024-03-06', 'fichierInsp...   

                                 documentsHorsInspection             date_maj  \
16     [{'identifiantFichier': 'xwOI57mkibOLwaTaqUaWl...  2024-09-21/05-30-29   
502                                                   []  2024-06-12/12-42-23   
2782   [{'identifiantFichier': '87af46b17c164f4abec88...  2024-10-21/18-18-00   
3675   [{'identifiantFichi

### Réinjection les codes manquants dans le dataframe complet :

In [9]:
data = data.join(filled[["newCodeInsee"]])
data["codeInsee"] = data["codeInsee"].combine_first(data["newCodeInsee"])
data = data.drop("newCodeInsee", axis=1)

### Evaluation des résultats

In [10]:
data.codeInsee.isnull().value_counts().to_frame("Nombre de codes manquants :")

Nombre de codes manquants :
codeInsee                             
False                            12364
True                                 1

### Données toujours manquantes

In [11]:
data[data.codeInsee.isnull()]

raisonSociale           adresse1 codePostal  \
8500  ATELIERS DES EPICES ET CONDIMENTS  19 RUE DE LA GARE      62360   

     codeInsee          commune  longitude  latitude  bovins  porcs  \
8500       NaN  PONT DE BRIQUES        NaN       NaN   False  False   

      volailles  ...  serviceAIOT    regime  rubriques  \
8500      False  ...    DREAL HdF  Non ICPE         []   

                             inspections documentsHorsInspection  \
8500  [{'dateInspection': '2015-01-12'}]                      []   

                 date_maj               adresse2 codeNaf           siret  \
8500  2024-12-24/15-00-00  SAINT ETIENNE AU MONT     NaN  53197729600023   

      adresse3  
8500       NaN  

[1 rows x 31 columns]

A date du 14/02/2025, une seule commune n'a pas été trouvée.

Effectivement, dans ce cas de figure, le lieu-dit (PONT DE BRIQUES) et la commune (SAINT ETIENNE AU MONT) ont été inversés : ceci explique que le score de la base adresse nationale n'ait pas été jugé suffisamment bon pour que le résultat de Saint-Etienne-au-Mont puisse être retenu...

Si cette fois, on décide d'utiliser l'API Nominatim en dernier recours, le code devient :

In [12]:
# On isole la(es) ligne(s) manquante(s)
missing = data[data.codeInsee.isnull()].copy()

# On concatène de nouveau les champs adresses :
cols = [f"adresse{x}" for x in range(1, 4)]
addresses = (
    missing[cols[0]]
    .str.cat(missing[cols[1:]], sep=" ", na_rep="")
    .str.replace(" +", " ", regex=True)
    .str.strip(" ")
)
missing.loc[:, "adresse"] = addresses

# Et on spécifie l'usage de Nominatim
missing = find_city(
    missing,
    year="last",
    x=False,
    y=False,
    dep=False,
    epsg=2154,
    city="commune",
    address="adresse",
    postcode="codePostal",
    use_nominatim_backend=True,
    field_output="newCodeInsee",
)

Utilisons pynsee pour récupérer la liste des communes pour contrôler le résultat (et pas simplement son code commune) :

In [13]:
# Présentation des résultats
from pynsee.localdata import get_area_list
cities = get_area_list("communes", date="*")
missing["newCodeInsee"].to_frame().merge(cities, left_on="newCodeInsee", right_on="CODE")

newCodeInsee   CODE                                                URI  \
0        62746  62746  http://id.insee.fr/geo/commune/adc106e4-5055-4...   

  AREA_TYPE DATE_CREATION            TITLE_SHORT DETERMINER_TYPE  \
0   Commune    1943-01-01  Saint-Étienne-au-Mont               0   

                   TITLE DATE_DELETION  
0  Saint-Étienne-au-Mont          None

💡 Nota : l'exécution de Nominatim ne conduit pas systématiquement au même résultat (et même parfois ne produit pas de résultat). Cela n'est pas totalement absurde, le hameau manquant étant à cheval sur plusieurs communes. Les résultats fournis restent  généralement pertinents.